pkg/auth/providers/azure/clients/client.go

```diff
@@ -0,0 +1,14 @@
+package clients
+
+// AzureClient specifies the subset of operations that a real client would delegate to some SDK for accessing
+// one of the two APIs to work with Active Directory resources - Azure AD Graph and Microsoft Graph.
+type AzureClient interface {
```

Comment:
A general note on this interface:

You have a lot of methods here which return `interface{}` objects. Generally speaking, I think it's best to avoid return types like this. It's difficult to consume when you call the function because you need to be aware about the internal implementation details of the function rather than just being able to accept/process a return type.

In several of these cases, I think that you can replace the `interface{}` with more specific objects (namely principals). You'll have to move a lot of the typecasting/conversion from the provider and into the client, but I think it's doable and would improve readability here substantially.